File: src/api/portfolio.ts
import type { KpiMetric, PortfolioData, SeriesPoint } from "@/lib/types";
import { borrowers, loans } from "@/data/dataset";
import { KES, pct } from "@/lib/format";
import { sleep } from "./client";

export const portfolioApi = {
  async getPortfolioData(): Promise<PortfolioData> {
    await sleep(200);

    const totalBorrowers = borrowers.length;
    const activeLoans = loans.filter((l) => l.status === "active").length;
    const totalExposure = loans
      .filter((l) => l.status === "active")
      .reduce((sum, l) => sum + l.amount, 0);
    const outstandingBalance = loans
      .filter((l) => l.status === "active")
      .reduce((sum, l) => sum + l.outstanding, 0);
    const avgPd =
      borrowers.reduce((sum, b) => sum + b.probabilityOfDefault, 0) / Math.max(1, borrowers.length);
    const defaultedLoans = loans.filter((l) => l.status === "defaulted").length;
    const defaultRate = defaultedLoans / Math.max(1, loans.length);

    const highRiskLoans = loans.filter((l) => {
      const b = borrowers.find((bor) => bor.id === l.borrowerId);
      return b?.riskBand === "high" || b?.riskBand === "severe";
    });
    const highRiskExposure = highRiskLoans.reduce((sum, l) => sum + l.outstanding, 0);

    const kpis: KpiMetric[] = [
      {
        key: "total_borrowers",
        label: "Total Borrowers",
        value: totalBorrowers.toString(),
        raw: totalBorrowers,
        change: 0.18,
        changeLabel: "+18% this quarter",
        intent: "positive",
        hint: "Total unique borrowers onboarded in the credit portfolio.",
      },
      {
        key: "active_loans",
        label: "Active Loans",
        value: activeLoans.toString(),
        raw: activeLoans,
        change: 0.12,
        changeLabel: "+12% active book",
        intent: "positive",
        hint: "Performing and active loan facilities currently disbursed.",
      },
      {
        key: "total_exposure",
        label: "Total Disbursed",
        value: KES(totalExposure, true),
        raw: totalExposure,
        change: 0.22,
        changeLabel: "+22% vs target",
        intent: "positive",
        hint: "Total original principal disbursed across active facilities.",
      },
      {
        key: "outstanding_balance",
        label: "Outstanding Balance",
        value: KES(outstandingBalance, true),
        raw: outstandingBalance,
        change: 0.14,
        changeLabel: "+14% active risk",
        intent: "neutral",
        hint: "Current outstanding unpaid principal awaiting maturity or repayment.",
      },
      {
        key: "average_pd",
        label: "Portfolio Mean PD",
        value: pct(avgPd, 1),
        raw: avgPd,
        change: -0.014,
        changeLabel: "-1.4pp risk reduction",
        intent: "positive",
        hint: "Weighted average probability of default across live facilities.",
      },
      {
        key: "default_rate",
        label: "Historical Default Rate",
        value: pct(defaultRate, 1),
        raw: defaultRate,
        change: -0.008,
        changeLabel: "-0.8pp vs annual limit",
        intent: "positive",
        hint: "Observed percentage of facilities transitioning into 90+ DPD default.",
      },
      {
        key: "high_risk_exposure",
        label: "High-Risk Exposure",
        value: KES(highRiskExposure, true),
        raw: highRiskExposure,
        change: 0.045,
        changeLabel: "+4.5% watch queue",
        intent: "negative",
        hint: "Capital deployed to High and Severe risk segments.",
      },
      {
        key: "repayment_rate",
        label: "On-Time Repayment",
        value: "92.4%",
        raw: 0.924,
        change: 0.021,
        changeLabel: "+2.1pp compliance",
        intent: "positive",
        hint: "Instalments received on or before contractual due date.",
      },
    ];

    const distribution = [
      {
        band: "low" as const,
        label: "Low Risk",
        count: 20,
        exposure: 4_250_000,
        share: 0.54,
      },
      {
        band: "medium" as const,
        label: "Medium Risk",
        count: 16,
        exposure: 2_150_000,
        share: 0.27,
      },
      {
        band: "high" as const,
        label: "High Risk",
        count: 10,
        exposure: 1_120_000,
        share: 0.14,
      },
      {
        band: "severe" as const,
        label: "Very High Risk",
        count: 6,
        exposure: 380_000,
        share: 0.05,
      },
    ];

    const defaultTrend: SeriesPoint[] = [
      { date: "Oct 25", rate: 5.2, target: 6.0 },
      { date: "Nov 25", rate: 4.9, target: 6.0 },
      { date: "Dec 25", rate: 5.4, target: 6.0 },
      { date: "Jan 26", rate: 4.8, target: 6.0 },
      { date: "Feb 26", rate: 4.4, target: 6.0 },
      { date: "Mar 26", rate: 4.1, target: 6.0 },
    ];

    const exposureTrend: SeriesPoint[] = [
      { date: "Oct 25", low: 3.1, medium: 1.8, high: 0.9, severe: 0.3 },
      { date: "Nov 25", low: 3.4, medium: 1.9, high: 0.9, severe: 0.3 },
      { date: "Dec 25", low: 3.7, medium: 2.0, high: 1.0, severe: 0.4 },
      { date: "Jan 26", low: 3.9, medium: 2.1, high: 1.0, severe: 0.3 },
      { date: "Feb 26", low: 4.1, medium: 2.1, high: 1.1, severe: 0.4 },
      { date: "Mar 26", low: 4.25, medium: 2.15, high: 1.12, severe: 0.38 },
    ];

    const approvalTrend: SeriesPoint[] = [
      { date: "Oct 25", approved: 72, volume: 220 },
      { date: "Nov 25", approved: 75, volume: 260 },
      { date: "Dec 25", approved: 70, volume: 310 },
      { date: "Jan 26", approved: 78, volume: 280 },
      { date: "Feb 26", approved: 80, volume: 340 },
      { date: "Mar 26", approved: 82, volume: 390 },
    ];

    const loanSizeDistribution = [
      { bucket: "< KES 25K", count: 18, exposure: 360_000 },
      { bucket: "25K–50K", count: 15, exposure: 580_000 },
      { bucket: "50K–100K", count: 12, exposure: 920_000 },
      { bucket: "100K–200K", count: 5, exposure: 850_000 },
      { bucket: "> 200K", count: 2, exposure: 550_000 },
    ];

    const defaultByPurpose = [
      { purpose: "Business Capital", defaultRate: 0.038, loans: 24 },
      { purpose: "School Fees", defaultRate: 0.021, loans: 16 },
      { purpose: "Agri Inputs", defaultRate: 0.062, loans: 14 },
      { purpose: "Asset Purchase", defaultRate: 0.029, loans: 10 },
      { purpose: "Medical", defaultRate: 0.048, loans: 8 },
      { purpose: "Emergency", defaultRate: 0.075, loans: 6 },
    ];

    const defaultByBand = [
      { band: "Low Risk", defaultRate: 0.012, loans: 32 },
      { band: "Medium Risk", defaultRate: 0.046, loans: 26 },
      { band: "High Risk", defaultRate: 0.148, loans: 14 },
      { band: "Very High Risk", defaultRate: 0.342, loans: 6 },
    ];

    const defaultByCohort = [
      { cohort: "Q1 2025", defaultRate: 0.058, borrowers: 38 },
      { cohort: "Q2 2025", defaultRate: 0.051, borrowers: 42 },
      { cohort: "Q3 2025", defaultRate: 0.046, borrowers: 48 },
      { cohort: "Q4 2025", defaultRate: 0.042, borrowers: 52 },
      { cohort: "Q1 2026", defaultRate: 0.039, borrowers: 56 },
    ];

    return {
      kpis,
      distribution,
      defaultTrend,
      exposureTrend,
      approvalTrend,
      loanSizeDistribution,
      defaultByPurpose,
      defaultByBand,
      defaultByCohort,
    };
  },
};

File: api/v1/routes/portfolio.py
"""
Portfolio risk, capital exposure distribution, and econometrics endpoints.
"""

from fastapi import APIRouter
from typing import List, Dict, Any
from pydantic import BaseModel
import json
import os

router = APIRouter()


def load_data(filename: str) -> List[dict]:
    path = os.path.join("data", filename)
    if os.path.exists(path):
        with open(path, "r") as f:
            return json.load(f)
    return []


@router.get("/summary")
async def get_portfolio_summary():
    borrowers = load_data("borrowers_synthetic.json")
    loans = load_data("loans_synthetic.json")

    total_borrowers = len(borrowers)
    active_loans = [l for l in loans if l.get("status") == "active"]
    total_exposure = sum(l.get("amount", 0) for l in active_loans)
    outstanding_bal = sum(l.get("outstanding", 0) for l in active_loans)
    defaulted_loans = [l for l in loans if l.get("status") == "defaulted"]

    avg_pd = (
        sum(b.get("probabilityOfDefault", 0) for b in borrowers) / max(1, total_borrowers)
    )
    default_rate = len(defaulted_loans) / max(1, len(loans))

    # Risk band distribution
    band_counts = {"low": 0, "medium": 0, "high": 0, "severe": 0}
    band_exposure = {"low": 0.0, "medium": 0.0, "high": 0.0, "severe": 0.0}

    for b in borrowers:
        band = b.get("riskBand", "medium")
        band_counts[band] = band_counts.get(band, 0) + 1
        band_exposure[band] = band_exposure.get(band, 0.0) + b.get("outstandingBalance", 0.0)

    distribution = [
        {
            "band": k,
            "label": f"{k.title()} Risk",
            "count": band_counts[k],
            "exposure": band_exposure[k],
            "share": round(band_counts[k] / max(1, total_borrowers), 3),
        }
        for k in ["low", "medium", "high", "severe"]
    ]

    return {
        "kpis": [
            {
                "key": "total_borrowers",
                "label": "Total Active Borrowers",
                "value": f"{total_borrowers:,}",
                "raw": total_borrowers,
                "change": 8.4,
                "changeLabel": "+8.4% MoM",
                "intent": "positive",
                "hint": "Number of unique verified borrowers with active profiles.",
            },
            {
                "key": "total_exposure",
                "label": "Total Active Exposure",
                "value": f"KES {total_exposure:,.0f}",
                "raw": total_exposure,
                "change": 14.2,
                "changeLabel": "+14.2% MoM",
                "intent": "positive",
                "hint": "Total principal committed across active facilities.",
            },
            {
                "key": "default_rate",
                "label": "Portfolio Default Rate",
                "value": f"{default_rate * 100:.1f}%",
                "raw": default_rate,
                "change": -0.6,
                "changeLabel": "-0.6% vs benchmark",
                "intent": "positive",
                "hint": "Observed 90+ DPD defaults over total originated loans.",
            },
            {
                "key": "avg_pd",
                "label": "Portfolio Mean PD",
                "value": f"{avg_pd * 100:.1f}%",
                "raw": avg_pd,
                "change": -0.4,
                "changeLabel": "Stable",
                "intent": "positive",
                "hint": "Calibrated probability of default across the loan book.",
            },
        ],
        "distribution": distribution,
        "defaultTrend": [
            {"date": "Q1 2025", "rate": 0.048, "benchmark": 0.065},
            {"date": "Q2 2025", "rate": 0.044, "benchmark": 0.064},
            {"date": "Q3 2025", "rate": 0.041, "benchmark": 0.062},
            {"date": "Q4 2025", "rate": 0.038, "benchmark": 0.060},
            {"date": "Q1 2026", "rate": 0.035, "benchmark": 0.058},
        ],
        "defaultByPurpose": [
            {"purpose": "Working Capital", "defaultRate": 0.038, "loans": 24},
            {"purpose": "Agriculture Inputs", "defaultRate": 0.082, "loans": 12},
            {"purpose": "Asset Purchase", "defaultRate": 0.029, "loans": 8},
            {"purpose": "School Fees", "defaultRate": 0.045, "loans": 5},
            {"purpose": "Emergency / Medical", "defaultRate": 0.095, "loans": 3},
        ],
    }
